src/foam/core/ClassLoader.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'ModelARequireExtension',
+  refines: 'foam.core.Model',
+
+  methods: [
+    function arequire() {
+      var X = this.__subContext__;
+      var promises = [];
+      if ( this.extends ) promises.push(X.arequire(this.extends));
+      for (var i = 0, a; a = this.axioms_[i]; i++) {
+        if ( a.arequire ) promises.push(a.arequire());
+      }
+      return Promise.all(promises);
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'RequiresARequireExtension',
+  refines: 'foam.core.Requires',
+
+  methods: [
+    function arequire() {
+      return this.__subContext__.arequire(this.path);
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'ClassLoader',
+
+  exports: [
+    'arequire',
+  ],
+
+  properties: [
+    {
+      name: 'pending',
+      class: 'Object',
+      factory: function() {
+        return {};
+      },
+    },
+  ],
+
+  methods: [
+    {
+      name: 'arequire',
+      class: 'foam.core.ContextMethod',
+      code: function(X, modelId) {
+        if (X.isRegistered(modelId)) return Promise.resolve();
+        if (this.pending[modelId]) return this.pending[modelId];
```

Comment:
Okay so this _would_ hang if you have a cycle, except that since you're registering the model before you call model.arequire(), this case is never hit via a circular dependency.

That being said, it feels more correct to me to do m.arequire() before we register the latch, as otherwise isRegistered() will return true for models that arn't quite loaded yet.  So I would make that change and then also fix this if to just return Promise.resolve()